### PR TITLE
Fix datapackage generation from multiple taxonomies

### PR DESCRIPTION
**Pull request by zschira**
# Overview
This PR fixes datapackage generation when using multiple taxonomies to extract XBRL filings. This fixes a metadata issue in the on-going 2023 FERC integration. It does this by merging multiple XBRL taxonomies into a single datapackage before parsing filings. This datapackage is used to structure the output DB and interpret filings during extraction.

### CODE REVIEW

File: src/ferc_xbrl_extractor/datapackage.py
Comment (zschira) [inline src/ferc_xbrl_extractor/datapackage.py:459]: Merging taxonomies is a bit weird, so I'll try to explain the logic a bit. `Resource`s in a datapackage correspond to a single table, so this will parse a single taxonomy normally, then for all subsequent taxonomies, it will look for any new or missing tables. For tables that existed in the previous version and the new version, it will call `merge_resources`, which then looks at the columns as well as the primary key of the table.

File: src/ferc_xbrl_extractor/datapackage.py
Comment (zschira) [inline src/ferc_xbrl_extractor/datapackage.py:337]: This is attempting to merge the schema of the same table from two different versions of a taxonomy. It first looks at primary key and raises an error if it's changed. Right now this never occurs, so it's not a problem, but if future versions change a primary key of a table, we'll need to decide how to handle this. After checking the primary key, it will look for any new or missing columns. If there are new columns, it will append these. Changes in the set of columns will lead to Nulls in the extracted table. So far there haven't been many changes in the column set, fortunately.

**Review by zschira — COMMENTED** [review]
The XBRL extraction is pretty janky, and this change is kind of adding some more jank on top, so I don't feel the best about it tbh. That being said, I waffled a lot on whether I should try to do something better, but ultimately feel like we just need to get 2023 data out the door, so maybe it's good enough for now.

File: src/ferc_xbrl_extractor/datapackage.py
Comment (zaneselvans) [reply src/ferc_xbrl_extractor/datapackage.py:337]: This is similar the actual behavior of the DBF derived data, in which we verified that as the DB schema evolved over the years, it was only ever additive -- new tables, and new columns within existing tables, but never the removal of a table of columns, which meant that we could always just use the latest published database to define a schema which was able to contain all years of data. But here you don't have that temporal dependency.

File: src/ferc_xbrl_extractor/datapackage.py
Comment (e-belfer) [inline src/ferc_xbrl_extractor/datapackage.py:346]: To clarify: does this mean we'll lose the metadata for older columns that have been dropped in the newer data?

File: src/ferc_xbrl_extractor/datapackage.py
Comment (e-belfer) [reply src/ferc_xbrl_extractor/datapackage.py:337]: Can we note this in the docstring of this method? Seems helpful to note what the underlying merge behavior is.

File: src/ferc_xbrl_extractor/datapackage.py
Comment (zschira) [reply src/ferc_xbrl_extractor/datapackage.py:346]: Yeah the older columns will remain, they're just logged so we can see what changed. I've added this to the docstring now.

File: src/ferc_xbrl_extractor/datapackage.py
Comment (zschira) [reply src/ferc_xbrl_extractor/datapackage.py:337]: @zaneselvans that makes sense, Dazhong and I have talked about making the extractor produce tidy tables, which would allow us to handle these changes in a much cleaner way, but decided to hold off for now.

@e-belfer yes! Sorry forgot to really go back and improve the documentation, but should be better now.

File: src/ferc_xbrl_extractor/datapackage.py
Comment (e-belfer) [inline src/ferc_xbrl_extractor/datapackage.py:348]: You're missing an extra space after `{self.name}` here.

**Review by e-belfer — CHANGES_REQUESTED** [review]
I successfully extracted the 2023 FERC XBRL data with this code change. 

One extremely minor line edit in this PR, which is my only actual blocking concern.

Just want to confirm for my sanity that recent XBRL changes break our ability to retroactively extract data from older Zenodo archives, and that this is an expected and intended decision?

Possibly out-of-scope: Are you expecting the datapackages to be valid frictionless datapackages (a la https://github.com/catalyst-cooperative/ferc-xbrl-extractor/issues/153)? I'm seeing tons of errors when I try to validate the one I generated. Again, just let me know if this one is out of scope.

One bigger concern: I _think_ that this update (or one before, or many before...) breaks the sample extraction code in the README. Which means that our instruction docs are now incorrect. We should definitely fix this. If it's quick to address, I'd prefer to do it here. If not, we should open an issue for this and make it a high priority after the 2023 data is merged.

File: src/ferc_xbrl_extractor/xbrl.py
Comment (e-belfer) [inline src/ferc_xbrl_extractor/xbrl.py:270]: Elsewhere we use `indent=n` to make our datapackages far more readable. Maybe worth considering here as the datapackage is currently hard to read.

Comment (e-belfer) [reply src/ferc_xbrl_extractor/xbrl.py:270]: See e.g., in our archiver repo:

new_datapackage.model_dump_json(by_alias=True, indent=4),
                    encoding="utf-8",

File: src/ferc_xbrl_extractor/datapackage.py
Comment (zschira) [reply src/ferc_xbrl_extractor/datapackage.py:348]: Fixed!

File: src/ferc_xbrl_extractor/xbrl.py
Comment (zschira) [reply src/ferc_xbrl_extractor/xbrl.py:270]: Good call, fixed!

**Comment by zschira** [comment]
> Possibly out-of-scope: Are you expecting the datapackages to be valid frictionless datapackages (a la https://github.com/catalyst-cooperative/ferc-xbrl-extractor/issues/153)? I'm seeing tons of errors when I try to validate the one I generated. Again, just let me know if this one is out of scope.

Hmm how were you checking the validation? We have an integration test, and validate the datapackage before writing to a json file.

> One bigger concern: I think that this update (or one before, or many before...) breaks the sample extraction code in the README. Which means that our instruction docs are now incorrect. We should definitely fix this. If it's quick to address, I'd prefer to do it here. If not, we should open an issue for this and make it a high priority after the 2023 data is merged.

Should be fixed now! I wonder if there's an easy way to test these in CI?

File: src/ferc_xbrl_extractor/datapackage.py
Comment (e-belfer) [inline src/ferc_xbrl_extractor/datapackage.py:464]:
```suggestion
        data from different years conforms to slightly different structures. This
```

File: README.rst
Comment (e-belfer) [inline README.rst:127]: @zschira are you able to run this successfully? I get:

```
 File "./ferc-xbrl-extractor/src/ferc_xbrl_extractor/instance.py", line 425, in instances_from_zip
 filing["filename"]: datetime.datetime.fromisoformat(
    ~~~~~~^^^^^^^^^^^^
TypeError: string indices must be integers, not 'str'
```

**Comment by e-belfer** [comment]
> Hmm how were you checking the validation? We have an integration test, and validate the datapackage before writing to a json file.

I just ran `frictionless validate datapackage.json` locally. I suspect this is a similar issue to the one in #153, which noted that our integration test wasn't catching this problem?

I get the following error for each table:
> Please provide "dialect.sql.table" for reading

**Review by e-belfer — APPROVED** [review]
I have a ~2~ outstanding questions about the `frictionless` issue ~and the README examples / our sample data~, but it's not clear to me if either are a new problem or not. Those outstanding, everything looks good to go to me. I'm going to ask that if you don't deal with them in this PR, we open new issues for both of them.

File: README.rst
Comment (zschira) [reply README.rst:127]: Did you do a `git pull`? I also had to update the example filings because they were the old format. When I run with the new filings everything works

Comment (e-belfer) [reply README.rst:127]: Oh silly, that did it. Sorry about that!

**Pull request merged**
